content/tutorials/reactjs-realtime-commenting.textile

```diff
@@ -0,0 +1,344 @@
+           <img src="https://bulma.io/images/placeholders/128x128.png" alt="Avatar" />
+          </p>
+        </figure>
+        <div className="media-content">
+          <div className="content">
+            <p>
+              <strong>{this.props.comment.name}</strong>
+              <br />
+              {this.props.comment.comment}
+            </p>
+          </div>
+        </div>
+      </article>
+    );
+  }
+}
+
+export default Comment;
+```
+
+This simply renders a single comment. It accepts the comment as props.
+
+h2. Step 11 - Create Comments Component
+
+Within the @components@ folder, create a new file named @Comments.js@ and add the code below into it:
+
+```[javascript]
+import React, { Component } from 'react';
+import Comment from './Comment';
+
+class Comments extends Component {
+  render() {
+    return (
+      <section className="section">
+        {
+          this.props.comments.map((comment, index) => {
+            return <Comment key={index} comment={comment} />
+          })
+        }
+      </section>
+    );
+  }
+}
+
+export default Comments;
+```
+
+This component accepts a @comments@ props and renders the Comment component for each of the comments available. It passes the actual comment to the Comment component as props.
+
+h2. Step 12 - Create App Component
+
+The App component will server as the parent component. Open @components/App.js@ and replace it with the code below:
+
+```[javascript]
+import React, { Component } from 'react';
+import CommentBox from './CommentBox';
+import Comments from './Comments';
+import Ably from '../ably';
+
+class App extends Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+        comments: []
+    }
+  }
+
+  render() {
+    return (
+      <section className="section">
+        <div className="container">
+          <div className="columns">
+            <div className="column is-half is-offset-one-quarter">
+              <CommentBox />
+              <Comments comments={this.state.comments} />
+            </div>
+          </div>
+        </div>
+      </section>
+    );
+  }
+}
+
+export default App;
+```
+
+This contains the components we created earlier. We define a @comments@ state which is an array of comments. It defaults to an empty array. This will be updated as we add comment through the comment form. Also, we pass the @comments@ state as props to the Comments component. This is how the Comments component get the comments it renders.
+
+h2. Step 13 - Subscribing to Comments
+
+We have seen how we can publish newly added comments. Now, we need to subscribe to those comments being published. We’ll do this within React lifecycle @componentDidMount()@ hook. Still within @components/App.js@, add the code below just before the @render()@ function:
+
+```[javascript]
+componentDidMount() {
+  // Subscribe to comments
+  const channel = Ably.channels.get('comments');
+
+  channel.subscribe('add_comment', (comment) => {
```

Comment:
Why is this logic in the App, when the publish logic is in the CommentBox?  Surely the CommentBox should handle publishing & subscribing?

Reply:
Because I want to update the `state` with the newly added comment. And `state` is only available at the parent/root component, hence the need to push it up to App component.

Reply:
But that's just wrong I am afraid as your logic is unnecessarily split simply because you want to work around state encapsulation in the component. You could simply expose a new function that allows a message to be added, and call that from the child, see https://stackoverflow.com/questions/35537229/how-to-update-parents-state-in-react.

Reply:
Thanks for pointing this out. This is exactly what I need initially, only to remove it because comments were been added twice due to the code I have in `componentDidMount`. I think I can revert to it since I can now use history.